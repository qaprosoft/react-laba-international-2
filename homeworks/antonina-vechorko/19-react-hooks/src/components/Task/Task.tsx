import React, {useCallback, useContext, useState, FC, ChangeEvent} from 'react';
import Input from '../Input/Input';
import IconButton from '../IconButton/IconButton';
import Checkbox from '../Checkbox/Checkbox';
import {TaskDispatchContext} from '../../context/Context';
import deleteIcon from '../../assets/delete.svg';
import editIcon from '../../assets/edit.svg';
import saveIcon from '../../assets/save.svg';

interface ITaskProps {
  task: {
    id: number;
    done: boolean;
    text: string;
  };
}

const Task: FC<ITaskProps> = ({task}) => {
  const dispatch = useContext(TaskDispatchContext);
  const [isEditing, setIsEditing] = useState(false);

  const handleEdit = () => {
    setIsEditing(true);
  };

  const handleSave = () => {
    setIsEditing(false);
  };

  const handleCheckboxClick = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      if (dispatch) {
        dispatch({
          type: 'toggle checkbox',
          task: {
            ...task,
            done: e.target.checked,
          },
        });
      }
    },
    [dispatch, task],
  );

  const handleEditText = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      if (dispatch) {
        dispatch({
          type: 'edit',
          task: {
            ...task,
            text: e.target.value,
          },
        });
      }
    },
    [dispatch, task],
  );

  const handleDeleteTask = (id: number) => {
    if (dispatch) {
      dispatch({
        type: 'delete',
        id: id,
      });
    }
  };

  return (
    <div className="task-list__item">
      <Checkbox checked={task.done} onChange={e => handleCheckboxClick(e)} />
      <Input
        value={task.text}
        onChange={e => handleEditText(e)}
        disabled={!isEditing}
        placeholder=""
      />
      <div className="action__wrapper">
        {isEditing ? (
          <IconButton src={saveIcon} alt="Save" onClick={handleSave} />
        ) : (
          <IconButton src={editIcon} alt="Edit" onClick={handleEdit} />
        )}
        <IconButton
          src={deleteIcon}
          alt="Delete"
          onClick={() => handleDeleteTask(task.id)}
        ></IconButton>
      </div>
    </div>
  );
};

export default Task;
