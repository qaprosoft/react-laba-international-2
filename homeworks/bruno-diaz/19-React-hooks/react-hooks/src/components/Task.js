import React, { useState } from "react";

export function Task(props) {
  const { task, onUpdateTask, onDeleteTask } = props;

  const [editing, setEditing] = useState(false);
  const [editedValue, setEditedValue] = useState(task.task);

  function handleEditClick() {
    setEditing(true);
  }

  function handleCancelEdit() {
    setEditing(false);
    setEditedValue(task.task);
  }

  function handleSaveEdit() {
    onUpdateTask({
      id: task.id,
      task: editedValue,
      completed: task.completed,
    });
    setEditing(false);
  }

  return (
    <>
      <div className="task-container" id={task.id}>
        {editing ? (
          <>
            <input
              className="input__form"
              type="text"
              value={editedValue}
              onChange={(e) => setEditedValue(e.target.value)}
            />
            <button onClick={handleSaveEdit} className="button__icon button__icon-update"></button>
            <button onClick={handleCancelEdit} className="button__icon button__icon-delete"></button>
          </>
        ) : (
          <>
            <span
              className={task.completed ? "input__form task-field completed" : "input__form"}
              onClick={() => onUpdateTask({ ...task, completed: !task.completed })}
            >
              {task.task}
            </span>
            <button onClick={handleEditClick} className="button__icon button__icon-update"></button>
            <button onClick={() => onDeleteTask(task.id)} className="button__icon button__icon-delete"></button>
          </>
        )}
      </div>
    </>
  );
}
