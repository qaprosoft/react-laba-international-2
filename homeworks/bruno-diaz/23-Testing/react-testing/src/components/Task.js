import React, { useState } from "react";

export function Task(props) {
  const { task, onUpdateTask, onDeleteTask } = props;

  const [editing, setEditing] = useState(false);
  const [editedValue, setEditedValue] = useState(task.task); 

  const [isCompleted, setIsCompleted] = useState(false);

  function EditionModeOn() {
    function handleChange(e) {
      const text = e.target.value;
      setEditedValue(text); 
    }

    function handleUpdate() {
      onUpdateTask({
        id: task.id,
        task: editedValue,
        completed: false,
      });
      setEditing(false);
    }

    return (
      <>
        <input
          className="input__form"
          type="text"
          value={editedValue} 
          onChange={handleChange}
        ></input>
        <button
          onClick={handleUpdate}
          className={"button__icon button__icon-update"}
        ></button>
        <button
          className={"button__icon button__icon-delete"}
          onClick={() => onDeleteTask(task.id)}
        ></button>
      </>
    );
  }

  function EditionModeOff() {
    return (
      <>
        <span
          className={isCompleted ? "input__form task-field completed" : "input__form"}
          onClick={() => setIsCompleted(!isCompleted)}
        >
          {task.task}
        </span>
        <button onClick={() => setEditing(true)} className={"button__icon button__icon-update"}></button>
        <button
          className={"button__icon button__icon-delete"}
          onClick={() => onDeleteTask(task.id)}
        ></button>
      </>
    );
  }

  return (
    <>
      <div className="task-container" id={task.id}>
        {editing ? <EditionModeOn /> : <EditionModeOff />}
      </div>
    </>
  );
}
