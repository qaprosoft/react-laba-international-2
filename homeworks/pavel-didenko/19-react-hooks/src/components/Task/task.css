.task {
  padding-left: 6.2rem;
}
