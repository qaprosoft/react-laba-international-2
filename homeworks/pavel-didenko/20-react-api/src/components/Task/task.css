.task {
  padding-left: 6.2rem;
}

.task__input:not(.task__input:disabled) {
  color: #000;
  font-size: 2.3rem;
}
