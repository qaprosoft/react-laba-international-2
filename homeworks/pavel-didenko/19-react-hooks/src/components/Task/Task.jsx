import React, {useRef, useState} from 'react';
import modifyIcon from '../../assets/img/icons/write.svg';
import removeIcon from '../../assets/img/icons/delete.svg';
import completeIcon from '../../assets/img/icons/task-complete.svg';
import './task.css';

const taskCompletedStyles = {
  textDecoration: 'line-through',
  color: 'green',
};

const Task = ({
  taskText,
  modifyTasks,
  index,
  removeTask,
  completed,
  setCompletedTask,
}) => {
  const [disabledModification, setDisabledModification] = useState(true);
  const [opacity, setOpacity] = useState(0);
  const [task, setTask] = useState(taskText);

  function modifyTaskHandler(newTaskText) {
    if (disabledModification) {
      setDisabledModification(false);
    } else {
      setDisabledModification(true);
      modifyTasks(index, newTaskText);
    }
  }

  function taskLengthValidator(task, minLength, maxLength) {
    if(task.length >= minLength && task.length <= maxLength) {
      modifyTaskHandler(task);
    } else {
      if(opacity === 0){
        setOpacity(1);
      }
    }
  }

  return (
    <div className="task" onClick={e => e.stopPropagation()}>
      <div className="task__inputs-wrapper">
        <input
          className="task__input"
          defaultValue={taskText}
          disabled={disabledModification}
          style={completed ? taskCompletedStyles : {}}
          onKeyUp={e => {
            if (e.key === 'Enter') {
              taskLengthValidator(task, 1, 33);
            }
          }}
          onChange={(e) => {
            setTask(e.target.value);
          }}
        ></input>
        <img
          className="task__icon"
          src={modifyIcon}
          alt="Modify task"
          onClick={() => taskLengthValidator(task, 1, 33)}
        />
        <img
          className="task__icon"
          src={removeIcon}
          alt="Remove task"
          onClick={() => removeTask(index)}
        />
        <img
          className="task__icon"
          src={completeIcon}
          alt="Complete icon"
          onClick={() => setCompletedTask(index)}
        />
      </div>
      <p className="task-creator__warning" style={{opacity: opacity}}>
        Task length must be from 1 to 33 characters
      </p>
    </div>
  );
};

export default Task;
