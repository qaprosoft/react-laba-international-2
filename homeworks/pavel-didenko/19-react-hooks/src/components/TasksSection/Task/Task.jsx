import React from 'react';
import Input from '../../Input/Input';
import modifyIcon from '../../../assets/img/icons/write.svg';
import removeIcon from '../../../assets/img/icons/delete.svg'
import './task.css';

const Task = () => {
  return (
    <div className="task">
      <Input className="task__input"/>
      <img className="task__icon" src={modifyIcon} alt="Modify task" />
      <img className="task__icon" src={removeIcon} alt="Remove task" />
    </div>
  );
}

export default Task;
