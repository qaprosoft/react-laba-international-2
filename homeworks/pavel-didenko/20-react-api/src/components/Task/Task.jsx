import React, {useState, useContext, useRef, useEffect} from 'react';
import modifyIcon from '../../assets/img/icons/write.svg';
import removeIcon from '../../assets/img/icons/delete.svg';
import completeIcon from '../../assets/img/icons/task-complete.svg';
import incompleteIcon from '../../assets/img/icons/task-incomplete.svg';
import './task.css';
import {MainContext} from '../../contexts/mainContext';
import useTaskEditValidator from '../../hooks/useTaskEditValidator';
import {taskLengthMessage} from '../../variables/errorMessages';
import IconButton from '../IconButton/IconButton';

const taskCompletedStyles = {
  textDecoration: 'line-through',
  color: 'green',
};

const Task = ({taskText, id, completed}) => {
  const [disabledModification, setDisabledModification] = useState(true);
  const {modifyTasks, removeTask, setCompletedTask} = useContext(MainContext);
  const [newTaskText, setNewTaskText] = useState(taskText);
  const {opacity, unique} = useTaskEditValidator(newTaskText, 1, 33);
  const taskInput = useRef(null);

  useEffect(() => {
    if (!disabledModification) taskInput.current.focus();
  }, [disabledModification]);

  function modifyTaskHandler(newTaskText) {
    if (disabledModification) {
      setDisabledModification(false);
    } else {
      setDisabledModification(true);
      modifyTasks(id, newTaskText);
    }
  }

  return (
    <div className="task" onClick={e => e.stopPropagation()}>
      <div className="task__inputs-wrapper">
        <input
          ref={taskInput}
          className="task__input"
          defaultValue={taskText}
          disabled={disabledModification}
          style={completed ? taskCompletedStyles : {}}
          onChange={e => {
            setNewTaskText(e.target.value);
          }}
          onKeyUp={e => {
            if (e.key === 'Enter' && opacity === 0) {
              modifyTaskHandler(newTaskText);
            }
          }}
        ></input>
        <IconButton
          src={modifyIcon}
          clickHandler={() => {
            if (opacity === 0 && unique) modifyTaskHandler(newTaskText);
          }}
          alt="Modify task"
        />
        <IconButton
          src={removeIcon}
          clickHandler={() => removeTask(id)}
          alt="Remove task"
        />
        <IconButton
          src={completed ? completeIcon : incompleteIcon}
          clickHandler={() => setCompletedTask(id)}
          alt="Complete icon"
        />
      </div>
      <p className="task-creator__warning" style={{opacity: opacity}}>
        {taskLengthMessage}
      </p>
    </div>
  );
};

export default Task;
