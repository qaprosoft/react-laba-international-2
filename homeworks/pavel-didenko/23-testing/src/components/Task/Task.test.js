import renderer from 'react-test-renderer';
import Task from './Task';

it('renders correctly', () => {
  const tree = renderer.create(<Task />).toJSON();
  expect(tree).toMatchSnapshot();
});
