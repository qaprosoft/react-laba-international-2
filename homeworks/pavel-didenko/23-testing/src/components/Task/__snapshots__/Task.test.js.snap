// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="task"
>
  <div
    className="task__inputs-wrapper"
  >
    <input
      className="task__input"
      disabled={true}
      onChange={[Function]}
      onKeyUp={[Function]}
      style={Object {}}
    />
    <button
      className="task__button"
    >
      <img
        alt="Modify task"
        className="task__icon"
        onClick={[Function]}
        src="write.svg"
      />
    </button>
    <button
      className="task__button"
    >
      <img
        alt="Remove task"
        className="task__icon"
        onClick={[Function]}
        src="delete.svg"
      />
    </button>
    <button
      className="task__button"
    >
      <img
        alt="Complete icon"
        className="task__icon"
        onClick={[Function]}
        src="task-incomplete.svg"
      />
    </button>
  </div>
  <p
    className="task-creator__warning"
    style={
      Object {
        "opacity": 0,
      }
    }
  >
    Task length must be from 1 to 33 characters
  </p>
</div>
`;
